File: src/codes/scattered-instaniation/py/rfc_factory.py
from abc import ABC, abstractmethod

class Logger:
    def log(self, message: str):
        print(f"[LOG]: {message}")

# Creator（抽象クラス）
class UserService(ABC):
    def __init__(self):
        self.logger = self.create_logger()

    @abstractmethod
    def create_logger(self) -> Logger:
        pass

    def create_user(self, name: str):
        self.logger.log(f"ユーザー作成: {name}")

# ConcreteCreator
class DefaultUserService(UserService):
    def create_logger(self) -> Logger:
        return Logger()

# 利用例
service = DefaultUserService()
service.create_user("Hiroshi")

File: src/codes/scattered-instaniation/ts/rfc_factory.ts
class Logger {
  log(message: string) {
    console.log(`[LOG]: ${message}`);
  }
}

// Creator（抽象クラス）
abstract class UserService {
  protected logger: Logger;

  constructor() {
    this.logger = this.createLogger(); // Factory Method を呼ぶ
  }

  protected abstract createLogger(): Logger;

  createUser(name: string) {
    this.logger.log(`ユーザー作成: ${name}`);
  }
}

// ConcreteCreator
class DefaultUserService extends UserService {
  protected createLogger(): Logger {
    return new Logger();
  }
}

// 利用例
const service = new DefaultUserService();
service.createUser("Hiroshi");
